pkg/csi/cinder/openstack/openstack_backups.go

```diff
@@ -0,0 +1,195 @@
+func (os *OpenStack) ListBackups(filters map[string]string) ([]backups.Backup, error) {
+	var allBackups []backups.Backup
+
+	// Build the Opts
+	opts := backups.ListOpts{}
+	for key, val := range filters {
+		switch key {
+		case "Status":
+			opts.Status = val
+		case "Name":
+			opts.Name = val
+		case "VolumeID":
+			opts.VolumeID = val
+		case "Marker":
+			opts.Marker = val
+		case "Limit":
+			opts.Limit, _ = strconv.Atoi(val)
+		default:
+			klog.V(3).Infof("Not a valid filter key %s", key)
+		}
+	}
+	mc := metrics.NewMetricContext("backup", "list")
+
+	allPages, err := backups.List(os.blockstorage, opts).AllPages()
+	if err != nil {
+		panic(err)
+	}
+	allBackups, err = backups.ExtractBackups(allPages)
+	if err != nil {
+		panic(err)
+	}
+
+	if mc.ObserveRequest(err) != nil {
+		return nil, err
+	}
+
+	return allBackups, nil
+}
+
+// DeleteBackup issues a request to delete the Backup with the specified ID from the Cinder backend.
+func (os *OpenStack) DeleteBackup(backupID string) error {
+	mc := metrics.NewMetricContext("backup", "delete")
+	err := backups.Delete(os.blockstorage, backupID).ExtractErr()
+	if mc.ObserveRequest(err) != nil {
+		klog.Errorf("Failed to delete backup: %v", err)
+	}
+	return err
+}
+
+// GetBackupByID returns backup details by id.
+func (os *OpenStack) GetBackupByID(backupID string) (*backups.Backup, error) {
+	mc := metrics.NewMetricContext("backup", "get")
+	backup, err := backups.Get(os.blockstorage, backupID).Extract()
+	if mc.ObserveRequest(err) != nil {
+		klog.Errorf("Failed to get backup: %v", err)
+		return nil, err
+	}
+	return backup, nil
+}
+
+// WaitBackupReady waits until backup is ready. It waits longer depending on
+// the size of the corresponding snapshot.
+func (os *OpenStack) WaitBackupReady(backupID string, snapshotSize int, backupMaxDurationSecondsPerGB int) error {
+	var err error
+
+	duration := time.Duration(backupMaxDurationSecondsPerGB*snapshotSize + 30)
+	ctx := context.Background()
+
+	err = os.waitBackupReadyWithContext(ctx, backupID, duration)
+	if wait.Interrupted(err) {
+		err = fmt.Errorf("timeout, Backup %s is still not Ready %v", backupID, err)
+	}
+
+	return err
+}
+
+// Supporting function for WaitBackupReady().
+// Allows for a timeout while waiting for the backup to be ready.
+func (os *OpenStack) waitBackupReadyWithContext(ctx context.Context, backupID string, duration time.Duration) error {
+	_, cancel := context.WithTimeout(context.Background(), duration*time.Second)
+	defer cancel()
+	var ok bool
+	var err error
+	for !ok {
+		ok, err = os.backupIsReady(backupID)
+		if err != nil {
+			return err
+		}
+		time.Sleep(7 * time.Second)
+	}
+	return err
+}
+
+// Supporting function for waitBackupReadyWithContext().
+// Returns true when the backup is ready.
+func (os *OpenStack) backupIsReady(backupID string) (bool, error) {
+	backup, err := os.GetBackupByID(backupID)
+	if err != nil {
+		return false, err
+	}
+
+	return backup.Status == backupReadyStatus, nil
```

Comment:
Yup! We do this in the `waitBackupReadyWithContext()` function, which calls this one (`backupIsReady()`). If this function returns an error, the error is passed to `waitBackupReadyWithContext()` and returned all the way up

Reply:
if os.GetBackupByID() returns any error, os.backupIsReady() returns it:
```
func (os *OpenStack) backupIsReady(backupID string) (bool, error) {
	backup, err := os.GetBackupByID(backupID)
	if err != nil {
		return false, err
	}

	return backup.Status == backupReadyStatus, nil
}
```
waitBackupReadyWithContext returns that same error:
```
	for {
		select {
		case <-ticker.C:
			done, err = os.backupIsReady(backupID)
			if err != nil {
				return err
			}

			if done {
				return nil
			}
		case <-ctx.Done():
			return ctx.Err()
		}
	}
```
Then WaitBackupReady receives that error and returns it:
```
	err = os.waitBackupReadyWithContext(backupID, duration)
	if err == context.DeadlineExceeded {
		err = fmt.Errorf("timeout, Backup %s is still not Ready: %v", backupID, err)
	}

	back, _ := os.GetBackupByID(backupID)

	if back != nil {
		return back.Status, err
	} else {
		return "Failed to get backup status", err
	}
```
This goes up CreateSnapshot()
```
		backup.Status, err = cs.Cloud.WaitBackupReady(backup.ID, snapSize, backupMaxDurationSecondsPerGB)
		if err != nil {
			klog.Errorf("Failed to WaitBackupReady: %v", err)
			return nil, status.Error(codes.Internal, fmt.Sprintf("CreateBackup failed with error %v", err))
		}
```

Reply:
I don't mean API returning an error. I mean `backup.Status` becoming `error`. This can happen and such state is irrecoverable. Here I caused it by specifying a non-existing AZ, but it can happen for various reasons, e.g. storage being full.

```
mdulko:~/ $ openstack volume backup create --availability-zone non-existing --snapshot test test 
+-----------+--------------------------------------+
| Field     | Value                                |
+-----------+--------------------------------------+
| id        | eb01d44f-5b1a-4e9e-b23e-7f979bf38dc1 |
| name      | None                                 |
| volume_id | d35ddf06-be8c-4205-861a-fae31c8d3600 |
+-----------+--------------------------------------+
mdulko:~/ $ openstack volume backup list
+--------------------------------------+------+-------------+--------+------+-------------+
| ID                                   | Name | Description | Status | Size | Incremental |
+--------------------------------------+------+-------------+--------+------+-------------+
| eb01d44f-5b1a-4e9e-b23e-7f979bf38dc1 | None | None        | error  |   10 | False       |
+--------------------------------------+------+-------------+--------+------+-------------+
```

Reply:
Nice, thanks!